File: AudioKit/Operations/Math/AKDifference.m
//
//  AKDifference.m
//  AudioKit
//
//  Created by Aurelius Prochazka on 2/21/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

#import "AKDifference.h"
#import "AKSum.h"

@implementation AKDifference {
    AKParameter *_minuend;
    AKParameter *_subtrahend;
}

- (instancetype)initWithInput:(AKParameter *)minuend minus:(AKParameter *)subtrahend
{
    self = [super initWithString:[self operationName]];
    
    if (self) {
        _minuend = minuend;
        _subtrahend = subtrahend;
        self.state = @"connectable";
        self.dependencies = @[_minuend, _subtrahend];
    }
    return self;
}

- (NSString *)inlineStringForCSD
{
    NSMutableString *inlineCSDString = [[NSMutableString alloc] init];
    
    [inlineCSDString appendString:@"akdifference("];
    [inlineCSDString appendString:[self inputsString]];
    [inlineCSDString appendString:@")"];
    
    return inlineCSDString;
}


- (NSString *)stringForCSD
{
    NSMutableString *csdString = [[NSMutableString alloc] init];
    
    [csdString appendFormat:@"%@ akdifference ", self];
    [csdString appendString:[self inputsString]];
    return csdString;
}

- (NSString *)inputsString {
    NSMutableString *inputsString = [[NSMutableString alloc] init];
    
    
    if ([_minuend class] == [AKAudio class]) {
        [inputsString appendFormat:@"%@, ", _minuend];
    } else {
        [inputsString appendFormat:@"AKAudio(%@), ", _minuend];
    }
    
    if ([_subtrahend class] == [AKAudio class]) {
        [inputsString appendFormat:@"%@", _subtrahend];
    } else {
        [inputsString appendFormat:@"AKAudio(%@)", _subtrahend];
    }

    return inputsString;
}

- (NSString *)udoString
{
    return @"\n"
    "opcode  akdifference, a, aa\n"
    "aMinuend, aSubtrahend xin\n"
    "aDifference sum aMinuend, -aSubtrahend\n"
    "xout aDifference\n"
    "endop\n";
}
@end

File: AudioKit/Operations/Math/AKDifference.swift
//
//  AKDifference.swift
//  AudioKit
//
//  Created by Aurelius Prochazka on 9/14/15.
//  Copyright © 2015 AudioKit. All rights reserved.
//

import Foundation

/** Difference of two inputs signals
*/
@objc class AKDifference : AKParameter {
    
    // MARK: - Properties
    
    /** Input to the mathematical function */
    private var minuend    = AKParameter()
    /** Input to the mathematical function */
    private var subtrahend = AKParameter()
    
    // MARK: - Initializers
    
    /** Instantiates the difference
    - parameter input: The input to be subtracted from.
    - parameter minus: The input that will be subtracted.
    */
    init(input minuendInput: AKParameter, minus subtrahendInput: AKParameter)
    {
        super.init()
        minuend = minuendInput
        subtrahend = subtrahendInput
        dependencies = [minuend, subtrahend]
    }
    
    /** Computation of the next value */
    override func compute() {
        leftOutput  = minuend.leftOutput  - subtrahend.leftOutput
        rightOutput = minuend.rightOutput - subtrahend.rightOutput
    }
}

/** Subtraction helper function */
func - (left: AKParameter, right: AKParameter) -> AKDifference {
    return AKDifference(input: left, minus: right)
}

/** Subtraction helper function */
func - (left: AKParameter, right: Float) -> AKDifference {
    return AKDifference(input: left, minus: akp(right))
}

/** Subtraction helper function */
func - (left: Float, right: AKParameter) -> AKDifference {
    return AKDifference(input: akp(left), minus: right)
}

/** Subtraction helper function */
func - (left: AKParameter, right: Int) -> AKDifference {
    return AKDifference(input: left, minus: akpi(right))
}

/** Subtraction helper function */
func - (left: Int, right: AKParameter) -> AKDifference {
    return AKDifference(input: akpi(left), minus: right)
}

